### PR TITLE
compiler: Optimize 'receive' across function calls

### PR DESCRIPTION
**Pull request by jhogberg**

### CODE REVIEW

**Review by bjorng — APPROVED** [review]
Overall looks good!

File: lib/compiler/src/beam_ssa_opt.erl [hunk 1]
```diff
@@ -1479,41 +1489,89 @@ do_opt_try([{L,Blk}|Bs]=Bs0, Ws0) ->
...
+        false ->
+            trim_try(Bs, UnreachablePads0, KilledTags0, [{L, Blk0} | Acc])
+    end;
+trim_try([{L, #b_blk{is=[#b_set{op=landingpad} | _]}=Blk}| Bs],
+         UnreachablePads0, KilledTags, Acc) ->
+    UnreachablePads = cerl_sets:add_element(L, UnreachablePads0),
+    trim_try(Bs, UnreachablePads, KilledTags, [{L, Blk} | Acc]);
+trim_try([{L, Blk} | Bs], UnreachablePads, KilledTags, Acc) ->
+    Successors = beam_ssa:successors(Blk),
+    Pads = maps:without(Successors, UnreachablePads),
```
Comment (bjorng) [inline lib/compiler/src/beam_ssa_opt.erl:1551]: What does Dialyzer think about this? (You are treating a cerl_set, which is opaque, as a map.)

Comment (jhogberg) [reply lib/compiler/src/beam_ssa_opt.erl:1551]: Fixed locally.

File: lib/compiler/src/beam_ssa.erl [hunk 2]
```diff
@@ -231,11 +230,17 @@ no_side_effect(#b_set{op=Op}) ->
...
+        _ ->
+            true
+    end;
 is_loop_header(#b_set{op=Op}) ->
     case Op of
         peek_message -> true;
-        wait -> true;
-        wait_timeout -> true;
         _ -> false
     end.
```
Comment (bjorng) [inline lib/compiler/src/beam_ssa.erl:330]:
```suggestion
    Op =:= peek_message.
```

File: lib/compiler/src/beam_ssa_type.erl [hunk 3]
```diff
@@ -1059,8 +1057,17 @@ will_succeed_1(#b_set{op=call}, _Src, _Ts, _Sub) ->
 will_succeed_1(#b_set{op=get_map_element}, _Src, _Ts, _Sub) ->
     maybe;
 
-will_succeed_1(#b_set{op=wait}, _Src, _Ts, _Sub) ->
-    no;
+will_succeed_1(#b_set{op=wait_timeout,args=[Timeout]}, _Src, Ts, _Sub) ->
+    case normalized_type(Timeout, Ts) of
+        #t_integer{elements={Min,Max}} when Min >= 0, Max =< 60000 ->
+            %% Timeouts that are too long can fail. Conservatively assume that
```
Comment (bjorng) [inline lib/compiler/src/beam_ssa_type.erl:1063]: Seems far too conservative me.

**Comment by garazdawi** [comment]
How hard would it be to add an option like +bin_opt_info that tells you if the receive optimization has triggered?

Should/Does the efficiency guide contain information about this?

**Comment by jhogberg** [comment]
It shouldn't be too difficult, I'll add it once I'm happy with the rest of the PR.

**Comment by jhogberg** [comment]
I've got a basic implementation working. The lack of human-readable variable names makes it a bit harder to interpret than it should be, but I'll spend a few hours trying to make it more readable.

```erlang
-module(t).
-export([t/0, q/0]).

t() ->
    %% t.erl:5: Warning: OPTIMIZED: created receive marker
    A = make_ref(),
    t_1(A).

t_1(A) ->
    %% t.erl:9: Warning: OPTIMIZED: used receive marker in 1st function argument
    receive
        A -> A
    end.

q() ->
    %% t.erl:14: Warning: OPTIMIZED: created receive marker
    {_, Ref} = spawn_monitor(fun() -> ok end),
    %% t.erl:15: Warning: OPTIMIZED: used receive marker created by erlang:spawn_monitor/1 at t.erl:14
    receive 
        Ref -> Ref
    end.
```

**Pull request merged**